stdlib/Random/src/XoshiroSimd.jl

```diff
@@ -0,0 +1,344 @@
+
+        code = """
+        %lshiftOp = shufflevector <1 x i64> <i64 17>, <1 x i64> undef, <$N x i32> zeroinitializer
+        %res = shl <$N x i64> %0, %lshiftOp
+        ret <$N x i64> %res
+        """
+        @eval @inline _shl17(x::NTuple{$N, VecElement{UInt64}}) = Core.Intrinsics.llvmcall($code,
+            NTuple{$N, VecElement{UInt64}},
+            Tuple{NTuple{$N, VecElement{UInt64}}},
+            x)
+
+        code = """
+        %lshiftOp = shufflevector <1 x i64> <i64 23>, <1 x i64> undef, <$N x i32> zeroinitializer
+        %rshiftOp = shufflevector <1 x i64> <i64 41>, <1 x i64> undef, <$N x i32> zeroinitializer
+        %lshifted = shl <$N x i64> %0, %lshiftOp
+        %rshifted = lshr <$N x i64> %0, %rshiftOp
+        %res = or <$N x i64> %lshifted, %rshifted
+        ret <$N x i64> %res
+        """
+        @eval @inline _rotl23(x::NTuple{$N, VecElement{UInt64}}) = Core.Intrinsics.llvmcall($code,
+            NTuple{$N, VecElement{UInt64}},
+            Tuple{NTuple{$N, VecElement{UInt64}}},
+            x)
+
+        code = """
+        %res = add <$N x i64> %1, %0
+        ret <$N x i64> %res
+        """
+        @eval @inline _plus(x::NTuple{$N, VecElement{UInt64}}, y::NTuple{$N, VecElement{UInt64}}) = Core.Intrinsics.llvmcall($code,
+            NTuple{$N, VecElement{UInt64}},
+            Tuple{NTuple{$N, VecElement{UInt64}}, NTuple{$N, VecElement{UInt64}}},
+            x, y)
+
+        code = """
+        %res = xor <$N x i64> %1, %0
+        ret <$N x i64> %res
+        """
+        @eval @inline _xor(x::NTuple{$N, VecElement{UInt64}}, y::NTuple{$N, VecElement{UInt64}}) = Core.Intrinsics.llvmcall($code,
+            NTuple{$N, VecElement{UInt64}},
+            Tuple{NTuple{$N, VecElement{UInt64}}, NTuple{$N, VecElement{UInt64}}},
+            x, y)
+
+        code = """
+        %res = and <$N x i64> %1, %0
+        ret <$N x i64> %res
+        """
+        @eval @inline _and(x::NTuple{$N, VecElement{UInt64}}, y::NTuple{$N, VecElement{UInt64}}) = Core.Intrinsics.llvmcall($code,
+            NTuple{$N, VecElement{UInt64}},
+            Tuple{NTuple{$N, VecElement{UInt64}}, NTuple{$N, VecElement{UInt64}}},
+            x, y)
+
+        code = """
+        %res = or <$N x i64> %1, %0
+        ret <$N x i64> %res
+        """
+        @eval @inline _or(x::NTuple{$N, VecElement{UInt64}}, y::NTuple{$N, VecElement{UInt64}}) = Core.Intrinsics.llvmcall($code,
+            NTuple{$N, VecElement{UInt64}},
+            Tuple{NTuple{$N, VecElement{UInt64}}, NTuple{$N, VecElement{UInt64}}},
+            x, y)
+
+        code = """
+        %tmp = insertelement <1 x i64> undef, i64 %1, i32 0
+        %shift = shufflevector <1 x i64> %tmp, <1 x i64> %tmp, <$N x i32> zeroinitializer
+        %res = lshr <$N x i64> %0, %shift
+        ret <$N x i64> %res
+        """
+        @eval @inline _lshr(x::NTuple{$N, VecElement{UInt64}}, y::Int64) = Core.Intrinsics.llvmcall($code,
+            NTuple{$N, VecElement{UInt64}},
```

Comment:
I think we could just use `ccall` for all of these and interpolate the `VecElement`s, like in my rotate example.

As an aside, I'd love it if Julia's codegen lowered `NTuple{N,VecElement{T}}`s with the same operations as `T`s. I assume this was already discussed somewhere (and rejected?).